### PR TITLE
ui-speedspacechart: add declivities layer in gev v2

### LINKED ISSUE #7731: osrd-ui: add declivities layer in GEV V2
### Description and goal

the GEV V2 should be able to display a linear representation of the declivities under the curve layer

### Acceptance criteria

- [ ] the declivities layer follows the design system mockup
- [ ] this layer is displayed or not in accordance to the settings panel
- [ ] this layer is synchronized with the user interactions (zoom, pan...)

### PR DESCRIPTION
**Pull request by RomainValls**
closes https://github.com/OpenRailAssociation/osrd/issues/7731

### CODE REVIEW

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+      const rectWidth = x2 - x1;
+
+      const rectHeight = (current.gradient / slopesRange) * availableHeight;
+
+      ctx.fillRect(x1, centerY, rectWidth, rectHeight);
+    }
+  } catch (error) {
+    console.error('Error while drawing declivity:', error);
+  }
```
Comment (kmer2016) [inline ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: Small perf. improvement here : avoid recalculating values that do not change in the loop. For `x1` and `x2`, we can store the second part of calculation in a variable just before the loop. (Not sure if `coef` is the right name for the variable)
```suggestion
  ctx.fillStyle = '#CFDDCE';

  try {
    const coef = ((width - CURVE_MARGIN_SIDES) / maxPosition) * ratioX +
        CURVE_MARGIN_SIDES / 2;
    
    for (let i = 0; i < slopes.length - 1; i++) {
      const current = slopes[i];
      const next = slopes[i + 1];

      const x1 = current.position * coef;
      const x2 = next.position * coef;

      const rectWidth = x2 - x1;
      const rectHeight = (current.gradient / slopesRange) * availableHeight;

      ctx.fillRect(x1, centerY, rectWidth, rectHeight);
    }
  } catch (error) {
    console.error('Error while drawing declivity:', error);
  }
```

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    return;
+  }
+
+  clearCanvas(ctx, width, height);
+
+  ctx.save();
+  ctx.translate(leftOffset, 0);
+
+  // Calculate total height available for drawing, excluding the margins
+  const availableHeight = height - MARGIN_TOP - MARGIN_BOTTOM - 30;
```
Comment (kmer2016) [inline ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: store [magic number](https://dev.to/ruben_alapont/magic-numbers-and-magic-strings-its-time-to-talk-about-it-ci2) in const to improve readability. For example, here, we can store `30` in a constant `MARGIN_OFFSET`

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  // Calculate the vertical center of the chart
+  const centerY = MARGIN_TOP + 30 + availableHeight / 2;
+
+  console.log('centerY:', centerY, 'maxGradient:', maxGradient);
+
+  ctx.fillStyle = '#CFDDCE';
```
Comment (kmer2016) [inline ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: store the color in a variable whose name indicates the element to which the color is to be applied. For example
`SLOPE_FILL_COLOR=#CFDDCE`

File: ui-speedspacechart/src/components/utils.ts [hunk 4]
```diff
@@ -13,8 +13,15 @@ type MaxPositionValues = {
...
-export const getGraphOffsets = (width: number, height: number) => {
-  const WIDTH_OFFSET = width - 60;
+type SlopesValues = {
+  minGradient: number;
+  maxGradient: number;
+  slopesRange: number;
+  maxPosition: number;
+};
+
+export const getGraphOffsets = (width: number, height: number, declivities: boolean) => {
```
Comment (kmer2016) [inline ui-speedspacechart/src/components/utils.ts]: Don't forget to update tests

File: ui-speedspacechart/src/components/utils.ts [hunk 5]
```diff
@@ -159,3 +166,15 @@ export const drawLinearLayerBackground = (
...
+ * Given a store including a list of slopes, return the position and value of min and max slopes
+ * @param store
+ */
+export const slopesValues = (store: Store): SlopesValues => {
+  const slopes = store.slopes;
+  const minGradient = Math.min(...slopes.map((data) => data.gradient));
+  const maxGradient = Math.max(...slopes.map((data) => data.gradient));
+  const slopesRange = maxGradient - minGradient;
+  const maxPosition = Math.max(...slopes.map((data) => data.position));
+  return { minGradient, maxGradient, slopesRange, maxPosition };
```
Comment (kmer2016) [inline ui-speedspacechart/src/components/utils.ts:208]: don't forget test

**Review by kmer2016 — COMMENTED** [review]
Great work ! Just some comments to improve readability. Generally, store magic number in const with right name. Same things with colors and move these constants at the top.

File: ui-speedspacechart/src/components/SpeedSpaceChart.tsx [hunk 6]
```diff
@@ -145,10 +148,11 @@ const SpeedSpaceChart = ({
           />
         </div>
       )}
+      <DeclivityLayer width={WIDTH_OFFSET} height={HEIGHT_OFFSET} store={store} />
       <CurveLayer width={WIDTH_OFFSET} height={HEIGHT_OFFSET} store={store} />
-      <AxisLayerY width={width} height={height} store={store} />
-      <MajorGridY width={width} height={height} store={store} />
-      <AxisLayerX width={width} height={height} store={store} />
+      <AxisLayerY width={declivities ? width - 42 : width} height={height} store={store} />
```
Comment (jacomyal) [inline ui-speedspacechart/src/components/SpeedSpaceChart.tsx]: This `declivities ? width - 42 : width` is too specific and custom to be repeated three times like this. The `42` should probably be a constant somewhere with a proper name, and the `declivities ? width - 42 : width` expression itself should be stored in a variable that describes what it represents.

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+const { CURVE_MARGIN_SIDES, MARGIN_TOP, MARGIN_BOTTOM } = MARGINS;
+
+export const drawDeclivity = (
+  ctx: CanvasRenderingContext2D,
+  width: number,
+  height: number,
+  store: Store
+) => {
+  const { slopes, ratioX, leftOffset } = store;
+  console.log('slopes:', { slopes, ratioX, leftOffset, width, height, store }, 'params and store');
```
Comment (jacomyal) [inline ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: (this should be removed before the merge)

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts (hunk 2 above)
Comment (jacomyal) [reply ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: Agreed, there are too many magic numbers in that PR.

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  ctx.save();
+  ctx.translate(leftOffset, 0);
+
+  // Calculate total height available for drawing, excluding the margins
+  const availableHeight = height - MARGIN_TOP - MARGIN_BOTTOM - 30;
+
+  // Calculate the vertical center of the chart
+  const centerY = MARGIN_TOP + 30 + availableHeight / 2;
+
+  console.log('centerY:', centerY, 'maxGradient:', maxGradient);
```
Comment (jacomyal) [inline ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: (same, should be removed before merging)

File: ui-speedspacechart/src/components/helpers/drawElements/declivity.ts (hunk 1 above)
Comment (RomainValls) [reply ui-speedspacechart/src/components/helpers/drawElements/declivity.ts]: Good point. I will let curve_margin_sides / 2 in the loop, as it's resulting in incorrect values when applied in the coef.

File: ui-speedspacechart/src/__tests__/utils.spec.ts [hunk 9]
```diff
@@ -94,6 +96,15 @@ describe('maxPositionValues', () => {
...
 });
 
+describe('slopesValues', () => {
+  it('should return the correct minGradient and maxGradient', () => {
+    const { minGradient, maxGradient, slopesRange } = slopesValues(store);
+    expect(minGradient).toBe(-10);
+    expect(maxGradient).toBe(50);
+    expect(slopesRange).toBe(60);
+  });
+});
```
Comment (kmer2016) [inline ui-speedspacechart/src/__tests__/utils.spec.ts]: Let's not forget to run the tests using npm run test to ensure we haven't broken anything. I have added one nominal case to the test.
```suggestion
  it('should return correct minGradient, maxGradient, slopesRange, and maxPosition', () => {
    const storeWithSlopes: Store = {
      ...store,
      slopes: [
        { gradient: 1, position: 10 },
        { gradient: 3, position: 20 },
        { gradient: 2, position: 15 },
        { gradient: 5, position: 25 },
      ],
    };

    const result = slopesValues(storeWithSlopes);

    expect(result).toEqual({
      minGradient: 1,
      maxGradient: 5,
      slopesRange: 4,
      maxPosition: 25,
    });
  });

  it('should handle empty slopes array', () => {
    const storeWithoutSlopes: Store = {
      ...store,
      slopes: [],
    };
    const result = slopesValues(storeWithoutSlopes);

    expect(result).toEqual({
      minGradient: Infinity,
      maxGradient: -Infinity,
      slopesRange: -Infinity,
      maxPosition: -Infinity,
    });
  });
```

File: ui-speedspacechart/src/__tests__/utils.spec.ts [hunk 10]
```diff
@@ -60,7 +61,8 @@ describe('getGraphOffsets', () => {
   it('should return the correct width and height offsets', () => {
     const width = 100;
     const height = 200;
-    const { WIDTH_OFFSET, HEIGHT_OFFSET } = getGraphOffsets(width, height);
+    const declivities = true;
+    const { WIDTH_OFFSET, HEIGHT_OFFSET } = getGraphOffsets(width, height, declivities);
     expect(WIDTH_OFFSET).toBe(40);
     expect(HEIGHT_OFFSET).toBe(120);
```
Comment (kmer2016) [inline ui-speedspacechart/src/__tests__/utils.spec.ts]: As I mentioned below, don't forget to run the tests locally using npm run test until we have CI set up to do it for us. I have added 3 cases : with declivities set to true, false and undefined.
```typescript
describe('getGraphOffsets', () => {
  const width = 200;
  const height = 150;

  it('should return correct width and height offsets when declivities is true', () => {
    const result = getGraphOffsets(width, height, true);

    expect(result).toEqual({
      WIDTH_OFFSET: 98,
      HEIGHT_OFFSET: 70,
    });
  });

  it('should return correct width and height offsets when declivities is false', () => {
    const result = getGraphOffsets(width, height, false);

    expect(result).toEqual({
      WIDTH_OFFSET: 140,
      HEIGHT_OFFSET: 70,
    });
  });

  it('should return correct width and height offsets when declivities is undefined', () => {
    const result = getGraphOffsets(width, height);

    expect(result).toEqual({
      WIDTH_OFFSET: 140,
      HEIGHT_OFFSET: 70,
    });
  });
});
```

File: ui-speedspacechart/src/__tests__/utils.spec.ts (hunk 9 above)
Comment (RomainValls) [reply ui-speedspacechart/src/__tests__/utils.spec.ts]: Thank you, done :)

File: ui-speedspacechart/src/__tests__/utils.spec.ts (hunk 10 above)
Comment (RomainValls) [reply ui-speedspacechart/src/__tests__/utils.spec.ts]: Done !

File: ui-speedspacechart/src/components/SpeedSpaceChart.tsx (hunk 6 above)
Comment (RomainValls) [reply ui-speedspacechart/src/components/SpeedSpaceChart.tsx]: Done :)

**Review by kmer2016 — APPROVED** [review]

**Pull request merged**